sql: implement drop function in legacy schema changer

chengxiong-ruan opened the pull request:
There are 5 commits:
(1) have function resolver return `ErrFunctionUndefined` error.
(2) implement drop function in legacy schema changer.
(3) support drop cascade of objects depended on by UDFs.
(4) disallow UDF usages from tables (injected the function resolver into declarative schema changer).
(5) disallow UDF usages from views and UDF.

chengxiong-ruan committed:
sql: wrap ErrFunctionUndefined for unknown functions

This commit adds a new ErrFunctionUndefined constant so
that callers of ResolveFunction can check if the returned
error is due to undefined functions.

Release note: None

chengxiong-ruan committed:
sql: implement DROP FUNCTION in legacy schema changer

This commit implements the DROP FUNCTION statement in legacy
schema changer.

Release note (sql change): DROP FUNCTION is implemented in
legacy schema changer. Now users can drop a function with
a function name or a function signature.

chengxiong-ruan committed:
sql: support drop cascade of objects depended on by UDF

This commits lift the failure in legacy schema changer when
dropping cascade objects which are depended on UDF. UDFs
are now dropped when these objects are dropped cascade.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85718)
<!-- Reviewable:end -->

chengxiong-ruan committed:
sql: disallow UDF usage from tables.

This commit adds a expression visitor which checks if there
is any UDF in the expression and sub-expression.
An unimplemented error is thrown if there's any UDF in any
table expression.

Release note: user-define functions are disallowed in any
expressions (column, index, constraint) in tables.

chengxiong-ruan committed:
sql: disallow udf usages in views or another udf

This commits adds a hack which set the function resolver to
nil in the context of CREATE VIEW and CREATE FUNcTION to let
it fall back to resolve only builtin functions. By doing, it's
the simplest way to disallow UDFs in views and function. To
make the error message less confusing, a hint is wrapped to
explain more.

Release note: None

ajwerner reviewed (commented):
This is looking good.

chengxiong-ruan commented on pkg/sql/drop_function.go:
```diff
@@ -12,24 +12,212 @@ package sql
+}
+
+func (p *planner) dropFunctionImpl(ctx context.Context, fnMutable *funcdesc.Mutable) error {
+	if fnMutable.Dropped() {
+		return errors.Errorf("function %q is already being dropped", fnMutable.Name)
+	}
```
oh I have dedupe logic with `fnResolved` to make sure only drop a function once, but good point that there should be a test for it. I'll add one.

chengxiong-ruan replied:
actually found a bug that the the ids were not added to the deduping set :( fixed that and added tests for it.

ajwerner replied:
We should also write a test that if you try to drop something twice without IF exists, that it fails

chengxiong-ruan replied:
added, thanks for pointing that out. actually we probably won't hit into the error of this check at all since the function is already removed from the schema descriptor, so later statements dropping the same function would never find the function at all.

chengxiong-ruan replied:
also added a test for the case where there are two same udf signatures in different schemas, and two statements in transaction to drop with the same signature, all two functions should be dropped, postgres does the same.

ajwerner commented on pkg/sql/drop_view.go:
```diff
@@ -138,39 +136,80 @@ func descInSlice(descID descpb.ID, td []toDelete) bool {
 }
 
-func (p *planner) canRemoveDependentView(
+func (p *planner) canRemoveDependentGeneric(
```
i don't see why this and canRemoveDependent both need to exist. just in-line this function into that one.

chengxiong-ruan replied:
oh, I don't like it as well, but there're use cases where the depended-on is not an table descriptor for example a `column`...

ajwerner replied:
okay, I see, can you use commentary to make the distinction clearer? Honestly, it feels like you should call the below one `canRemoveDependentFromTable` and just call this one `canRemoveDependent`?

chengxiong-ruan replied:
good idea. changed.

ajwerner commented on pkg/sql/opt/optbuilder/create_function.go:
```diff
@@ -47,6 +52,25 @@ func (b *Builder) buildCreateFunction(cf *tree.CreateFunction, inScope *scope) (
+				panic(
+					errors.WithHint(
+						recErr,
+						"There is probably a typo in function name. Or the intention was to use a user-defined "+
```
this probably is funny

chengxiong-ruan replied:
:(

ajwerner commented on pkg/sql/opt/optbuilder/create_function.go:
```diff
@@ -47,6 +52,25 @@ func (b *Builder) buildCreateFunction(cf *tree.CreateFunction, inScope *scope) (
 		b.qualifyDataSourceNamesInAST = false
+
+		b.semaCtx.FunctionResolver = preFuncResolver
+		switch recErr := recover().(type) {
```
extract a helper give you have this twice

chengxiong-ruan replied:
good point, done.

ajwerner commented on pkg/sql/catalog/descs/direct.go:
```diff
@@ -46,6 +46,12 @@ type Direct interface {
 
+	// MustGetDescByID looks up a descriptor (of any type) given its ID, returning
+	// an error if the descriptor is not found.
+	MustGetDescByID(
```
i didn't see this get used. what's this for?

chengxiong-ruan replied:
This is used in `dependentError` function to get a descriptor then do `switch` on descriptor type there.

ajwerner replied:
Can we not plumb in a collection?

chengxiong-ruan replied:
👍  dropped this method and use collection instead.

ajwerner reviewed (approved):
LGTM

chengxiong-ruan commented:
TFTR!
bors r=ajwerner
